script/modelling/space_shuttle_reentry_trajectory.jl

```diff
@@ -0,0 +1,281 @@
+#' $$
+
+#' As explained in the book, our goal is to maximize the final crossrange,
+#' which is equivalent to maximizing the final latitude of the vehicle,
+#' i.e., $J = \theta(t_F)$.
+
+#' ![Max crossrange shuttle reentry](https://i.imgur.com/1tUN3wM.png)
+
+#' ### Approach 
+#' We will use a discretized model of time, with a fixed number of discretized points, $n$.
+#' The decision variables at each point are going to be the state of the vehicle and the controls commanded to it.
+#' In addition, we will also make each time step size $\Delta t$ a decision variable;
+#' that way, we can either fix the time step size easily,
+#' or allow the solver to fine-tune the duration between each adjacent pair of points.
+#' Finally, in order to approximate the derivatives of the problem dynamics, we will use either rectangular
+#' or [trapezoidal](http://en.wikipedia.org/wiki/Trapezoidal_rule) integration.
+
+#' ### Disclaimer
+#' Do not try to actually land a Space Shuttle using this notebook! 😛
+#' There's no mesh refinement going on, which can lead to unrealistic trajectories having position
+#' and velocity errors with orders of magnitude $10^4$ ft and $10^2$ ft/sec, respectively.
+
+using Interpolations, Plots
+using JuMP, Ipopt  # , KNITRO
+
+#+
+
+# Global variables
+const w  = 203000  # weight (lb)
+const g₀ = 32.174  # acceleration (ft/sec^2)
+const m  = w / g₀  # mass (slug)
+
+# Aerodynamic and atmospheric forces on the vehicle
+const ρ₀ =  0.002378
+const hᵣ =  23800.0
+const Rₑ =  20902900.0
+const μ  =  0.14076539e17
+const S  =  2690.0
+const a₀ = -0.20704
+const a₁ =  0.029244
+const b₀ =  0.07854
+const b₁ = -0.61592e-2
+const b₂ =  0.621408e-3
+const c₀ =  1.0672181
+const c₁ = -0.19213774e-1
+const c₂ =  0.21286289e-3
+const c₃ = -0.10117249e-5
+
+#+
+
+user_options_ipopt = (
+    "mu_strategy" => "monotone",
+    "linear_solver" => "ma27",
+)
+
+# user_options_knitro = (
+#     "algorithm" => KNITRO.KN_ALG_BAR_DIRECT,
+#     "bar_murule" => KNITRO.KN_BAR_MURULE_QUALITY,
+#     "linsolver" => KNITRO.KN_LINSOLVER_MA27,
+# )
+
+# Create JuMP model, using Ipopt as the solver
+model = Model(optimizer_with_attributes(Ipopt.Optimizer, user_options_ipopt...))
+# model = Model(optimizer_with_attributes(KNITRO.Optimizer, user_options_knitro...))
+
+n = 2009  # number of mesh points (knots)
+
+@variables(model, begin
+               0 ≤ scaled_h[1:n]                # altitude (ft) / 1e5
+                          ϕ[1:n]                # longitude (rad)
+    deg2rad(-89) ≤        θ[1:n] ≤ deg2rad(89)  # latitude (rad)
+            1e-4 ≤ scaled_v[1:n]                # velocity (ft/sec) / 1e4
+    deg2rad(-89) ≤        γ[1:n] ≤ deg2rad(89)  # flight path angle (rad)
+                          ψ[1:n]                # azimuth (rad)
+    deg2rad(-90) ≤        α[1:n] ≤ deg2rad(90)  # angle of attack (rad)
+    deg2rad(-89) ≤        β[1:n] ≤ deg2rad( 1)  # banck angle (rad)
+    #        0.5 ≤       Δt[1:n] ≤ 1.5          # time step (sec)
```

Comment:
Why is this line commented out?

Reply:
Good question. There are two lines in the source code for specifying time:
```julia
    #        0.5 ≤       Δt[1:n] ≤ 1.5          # time step (sec)
                         Δt[1:n] == 1.0         # time step (sec)
```
The first one (currently commented) allows some wiggle room for the solver to adjust the time step size between pairs of mesh points. This is neat because it allows the solver to figure out which parts of the flight require more dense discretization than others. (Remember, the number of discretised points is fixed, and this example does not implement mesh refinement.) However, this is makes the problem more complex to solve, and therefore leads to a longer computation time.

The second line (the one not commented) fixes the duration of every time step to exactly 1.0 seconds. This allows the problem to be solved faster. The solver is still able to find a solution to the problem because I "cheated": I know that the close-to-optimal total duration of the flight is ~33 minutes, so I set the number of knots to `n = 2009` for the fixed time step to be set to 1.0 seconds.